Sources/TSCBasic/Downloader.swift

```diff
@@ -0,0 +1,136 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+/// Error enumeration that can be returned by a `Downloader` conforming type.
+public enum DownloaderError: Error {
+
+    /// Error thrown when the downloader fails to establish a connection to the server.
+    case clientError(error: Error)
+
+    /// Error thrown when the downloader received an invalid status code from the server.
+    case serverError(statusCode: Int)
+}
+
+/// The `Downloader` protocol abstract away the download of a file with a progress report.
+public protocol Downloader {
+
+    /// Downloads a file and keeps the caller updated on the progress and completion.
+    ///
+    /// - Parameters:
+    ///   - url: The `URL` to the file to download.
+    ///   - progress: A closure to receive the download's progress as a fractional value between `0.0` and `1.0`.
+    ///   - completion: A closure to be notifed of the completion of the download as a `Result` type containing the
+    ///     temporary `AbsolutePath` to the downloaded file on success, or the `DownloaderError` encountered on failure.
+    func downloadFile(
```

Comment:
Where does it download the file? Shouldn't that be an input?

Reply:
In my initial version of this proposal, it was. But as we want to use this to download archives and unzip them afterwards, I thought it would require an unnecessary move. But I’m totally ready to allow choosing the destination again. What do you think?

Reply:
Ah, looks like URLSession automatically downloads the file in some temporary location. Seems fine but how does it interact with:

> You must move this file or open it for reading before your completion handler returns. Otherwise, the file is deleted, and the data is lost.

Reply:
Agreed. I modified the downloader to move the file to a destination path.